[8.x] Resolve component class with $attributes property

panda-madness opened the pull request:
This small change allows one to type hint the $attributes parameter in the class constructor. This parameter contains an array of all attributes that were passed to this component.

```blade
<x-test-subject :foo="'bar'" baz="cookies" />
```
```php
<?php

namespace App\View\Components;

class TestSubject extends Base
{
    public function __construct($attributes)
    {
        dump($attributes); // contains ['foo' => 'bar', 'baz' => 'cookies']
    }
}

```

As this change reserves the `$attribute` parameter name it's technically breaking, although it's pretty small, so not sure whether this is major or minor feature.

I will adjust the tests if this PR gets a green light.

<!--
Please only send a pull request to branches which are currently supported: https://laravel.com/docs/releases#support-policy 

If you are unsure which branch your pull request should be sent to, please read: https://laravel.com/docs/contributions#which-branch

Pull requests without a descriptive title, thorough description, or tests will be closed.

In addition, please describe the benefit to end users; the reasons it does not break any existing features; how it makes building web applications easier, etc.
-->

panda-madness committed:
Resolve component class with $attributes property

This small change allows one to type hint the $attributes parameter in the class constructor. This parameter contains an array of all attributes that were passed to this component.

GrahamCampbell commented:
Thanks for the PR, however the tests are failing.

panda-madness committed:
Fix tests

panda-madness commented:
@GrahamCampbell I've fixed the tests.
Also wanted to ask if this is something that can be merged in 7.x instead?

driesvints commented:
@panda-madness please revert all the unnecessary indentations.

driesvints commented on tests/View/Blade/BladeComponentTagCompilerTest.php:
```diff
@@ -32,30 +32,30 @@ public function testBasicComponentParsing()
         $result = (new ComponentTagCompiler(['alert' => TestAlertComponent::class]))->compileTags('<div><x-alert type="foo" limit="5" @click="foo" required /><x-alert /></div>');
 
-        $this->assertSame("<div> @component('Illuminate\Tests\View\Blade\TestAlertComponent', [])
+        $this->assertSame("<div> @component('Illuminate\Tests\View\Blade\TestAlertComponent', ['attributes' => ['type' => 'foo','limit' => '5','@click' => 'foo','required' => true]])
```
Add a proper space after each array ellement:

```suggestion
        $this->assertSame("<div> @component('Illuminate\Tests\View\Blade\TestAlertComponent', ['attributes' => ['type' => 'foo', 'limit' => '5', '@click' => 'foo', 'required' => true]])
```

Please do this for all changes below.

panda-madness replied:
I've had to modify the existing function [here](https://github.com/laravel/framework/blob/7.x/src/Illuminate/View/Compilers/ComponentTagCompiler.php#L390) to make this happen. I've adjusted the tests accordingly.

panda-madness replied:
[This](https://github.com/laravel/framework/pull/32505/files#diff-6a6ea7a2d1468b4ad89232932195b939R400) is the relevant change.

driesvints commented on tests/View/Blade/BladeComponentTagCompilerTest.php:
```diff
@@ -32,30 +32,30 @@ public function testBasicComponentParsing()
 <?php \$component->withName('alert'); ?>
 <?php \$component->withAttributes(['type' => 'foo','limit' => '5','@click' => 'foo','required' => true]); ?>\n".
-"@endcomponentClass  @component('Illuminate\Tests\View\Blade\TestAlertComponent', [])
+            "@endcomponentClass  @component('Illuminate\Tests\View\Blade\TestAlertComponent', ['attributes' => []])
```
Please revert all of these unnecessary indentations.

panda-madness replied:
Done! Not sure what happened here, PHPStorm messed this up a bit.

panda-madness committed:
Fix tests, add spaces to array elements

taylorotwell commented:
You can access $this->attributes from within the `render` method.

panda-madness commented:
@taylorotwell [here](https://github.com/panda-madness/laravel-component-attributes) is a reproducible demo showing that that is not the case, `$this->attributes` is still null even in `render`. Note the `TestSubject` component.

<img width="438" alt="Screen Shot 2020-04-23 at 20 49 54" src="https://user-images.githubusercontent.com/6180087/80113368-08621d80-85a4-11ea-8c35-b9c8d03a3cb9.png">

driesvints commented:
@panda-madness you commited your custom changes together with the framework skeleton which makes it super hard to find out what's wrong.

panda-madness commented:
@driesvints Sorry about that. I've added some changes in a single [commit](https://github.com/panda-madness/laravel-component-attributes/commit/853f451f6b03787255aa4807cb8d0f23f74c75ca) to mark every file I modified, I'll know to separate my changes in the future.

panda-madness commented:
Bump in the night, @taylorotwell

driesvints commented:
@panda-madness I've actually stumbled onto this myself in the meantime. I'll have a look at this today.

panda-madness commented:
@driesvints something related to blade svg I guess? :)

driesvints commented:
@panda-madness heh yeah.

I discovered that you can still use attributes in your render method like this actually:

```php
    public function render()
    {
        return <<<'blade'
        <div class="<div {{ $attributes }}">
            Foo
        </div>
blade;
    }
```

Is there a special reason why you need to access the attributes? It's best that you provide a real-life use case. I'm having a look tomorrow again so let me know.

panda-madness commented:
@driesvints I was trying to make a sort of dynamic component that would render other components and pass along the attributes that it receives, sort of like in [Vue](https://vuejs.org/v2/guide/components.html#Dynamic-Components). For that I needed to extract 1 defined attribute (the component name), and pass everything else in the template.

driesvints commented:
@panda-madness I'm looking into doing that exact thing but you have to be aware that `$data != $attributes`. Data is all the data from the view piped through to components and attributes are the attributes defined in the html tag of the component.

What I would do in your case is dynamically render different views which have these components defined. Then you don't need to define them in your render method.

driesvints commented:
@panda-madness I've looked further into this but was unable to find a clean solution. I think it's a no-go on attributes being available in the render method for now.

panda-madness commented:
@driesvints would you consider petitioning Taylor to re-open this PR then? I suspect he might have unsubscribed from the discussion here.

driesvints commented:
No, because I don't think this is the right solution either to mix data with attributes.

panda-madness commented:
@driesvints I might be greatly misunderstanding how component tag compilation works, so please correct me if I'm wrong.

From what I'm seeing, `$data` initially contains all of the passed HTML attributes, and nothing else. You can see [here](https://github.com/panda-madness/framework/blob/6d7aa4f6024731016c303148f3ee7e2b789dfcfc/src/Illuminate/View/Compilers/ComponentTagCompiler.php#L114) and [here](https://github.com/panda-madness/framework/blob/6d7aa4f6024731016c303148f3ee7e2b789dfcfc/src/Illuminate/View/Compilers/ComponentTagCompiler.php#L158) where the attributes are Regex extracted and passed to `$this->componentString()`. The partitioning afterwards is, I assume, made to specifically enable [constructor dependency injection](https://github.com/laravel/framework/blob/7.x/src/Illuminate/View/Compilers/Concerns/CompilesComponents.php#L65) via the service container, since the container needs named arguments for that.

I don't really see where we're mixing anything that's not supposed to be mixed, as far as I can see we are only operating on the attributes that are passed to the component tag.

Again, if I'm mistaken I'm sorry to have taken so much of your time on this.

driesvints commented:
Technically your change will work momentarily I believe but I think it'll collide with changes made in https://github.com/laravel/framework/pull/32576 for the next major release.

panda-madness commented:
Is there a way to trigger the test suite against the new master branch?

driesvints commented:
Yeah, run them locally.

bastinald commented:
This should have been merged.

Attributes and slots should be accessible through the class constructor.

Here is a use case example for a component using Livewire:

```php
class Input extends Component
{
    public $model;

    public function __construct()
    {
        $this->model = $this->attributes->whereStartsWith('wire:model')->first();
    }

    public function render()
    {
        return <<<'HTML'
            <div class="mb-4 space-y-1">
                <input {{ 
                    $attributes->class([
                        'w-full placeholder-gray-400 rounded-lg',
                        'border-gray-400' => !$errors->has($model),
                        'border-red-600' => $errors->has($model),
                    ])->merge([
                        'type' => 'text',
                    ])
                }}>

                @error($model)
                    <p class="text-sm text-red-600">{{ $message }}</p>
                @enderror
            </div>
        HTML;
    }
}

```
